### PR TITLE
remove ability to give a git revision to devved packages

### PR DESCRIPTION
**Pull request by KristofferC**
Doing `dev Foo#branch1` and then `dev Foo#branch2` and then not get a change of branch is confusing. For now, just allow downloading the repo at the main branch and then one can just use git to get the repo in whatever state is desired.

### CODE REVIEW

**Comment by KristofferC** [comment]
I don't really understand how this test error happens due to this PR

```
Generating project Foo:
    Foo/Project.toml
    Foo/src/Foo.jl
(pkg.repo).url = "Foo"
abspath((pkg.repo).url) = "/private/var/folders/64/76tk_g152sg6c6t0b4nkn1vw0000gn/T/tmppjyCtY/HelloWorld/Foo"
realpath(abspath((pkg.repo).url)) = "/private/var/folders/64/76tk_g152sg6c6t0b4nkn1vw0000gn/T/tmppjyCtY/HelloWorld/Foo"
(ctx.env).project_file = "/var/folders/64/76tk_g152sg6c6t0b4nkn1vw0000gn/T/tmpFTWtM6/environments/v1.0/Project.toml"
ERROR: SystemError: realpath: No such file or directory
Stacktrace:
 [1] #systemerror#39(::Nothing, ::Function, ::Symbol, ::Bool) at ./error.jl:106
 [2] systemerror at ./error.jl:106 [inlined]
 [3] realpath(::String) at ./path.jl:322
 [4] macro expansion at ./show.jl:556 [inlined]
```

**Comment by fredrikekre** [comment]
I got that error too in some of my PRs, really strange...

**Comment by fredrikekre** [comment]
Wait, did you add those `@show`s for this PR? They look like the ones I added when debugging https://github.com/JuliaLang/Pkg.jl/pull/490

**Comment by KristofferC** [comment]
I added them locally.

**Comment by 00vareladavid** [comment]
It seems isolating that test allows it to pass... maybe its interacting with the one above it?

**Comment by fredrikekre** [comment]
I think we can safely change https://github.com/JuliaLang/Pkg.jl/blob/d2b8419063861e2a2315686a0db332603b2871e9/src/Types.jl#L248 to
```julia
project_file = joinpath(realpath(env), Base.project_names[end]) 
```
and
https://github.com/JuliaLang/Pkg.jl/blob/d2b8419063861e2a2315686a0db332603b2871e9/src/Types.jl#L250-L251
to
```julia
project_file = endswith(env, ".toml") ? abspath(realpath(env)) : 
    abspath(realpath(env), Base.project_names[end]) 
```
?

File: src/Types.jl [hunk 1]
```diff
@@ -511,7 +511,9 @@ function handle_repos_develop!(ctx::Context, pkgs::AbstractVector{PackageSpec},
                     # Relative paths are given relative pwd() so we
                     # translate that to be relative the project instead.
                     # `realpath` is needed to expand symlinks before taking the relative path.
-                    pkg.path = relpath(realpath(abspath(pkg.repo.url)), realpath(dirname(ctx.env.project_file)))
+                    project_dir = dirname(ctx.env.project_file)
+                    isdir(project_dir) && (project_dir = realpath(project_dir))
```
Comment (fredrikekre) [inline src/Types.jl]: This needs to call `realpath` on the part of `ctx.env.project_file` which exists, i.e. given
`/a/b/c/Project.toml`, if only `/a/b` exists, we need something like `joinpath(realpath(/a/b), c/Project.toml)`.

File: src/Types.jl [hunk 2]
```diff
@@ -511,7 +511,9 @@ function handle_repos_develop!(ctx::Context, pkgs::AbstractVector{PackageSpec},
                     # Relative paths are given relative pwd() so we
```
Comment (fredrikekre) [inline src/Types.jl]: We can also change this, such that when deving a path you write the path relative the project file, but that will be very annoying I think.

File: src/Types.jl (hunk 1 above)
Comment (fredrikekre) [reply src/Types.jl]: Basically we need a `safe_realpath`, somehting like
```
julia> function safe_realpath(path)
           ispath(path) && return realpath(path)
           a, b = splitdir(path)
           return joinpath(safe_realpath(a), b)
       end
safe_realpath (generic function with 1 method)

julia> str = "/home/fredrik/Foo/test/Project.toml"
"/home/fredrik/Foo/test/Project.toml"

julia> realpath(str)
ERROR: SystemError: realpath: No such file or directory
Stacktrace:
 [1] #systemerror#39(::Nothing, ::Function, ::Symbol, ::Bool) at ./error.jl:106
 [2] systemerror at ./error.jl:106 [inlined]
 [3] realpath(::String) at ./path.jl:322
 [4] top-level scope at none:0

julia> safe_realpath(str)
"/home/fredrik/Foo/test/Project.toml"
```

File: src/Types.jl (hunk 2 above)
Comment (KristofferC) [reply src/Types.jl]: Yes

**Comment by 00vareladavid** [comment]
Well that was demoralizing... :sob: . Maybe I'm missing something: is it really `realpath` that is failing? Wouldn't that be a Julia bug?

**Comment by fredrikekre** [comment]
:tada:

**Pull request merged**